src/machi_cr_client.erl

```diff
@@ -548,12 +548,119 @@ do_read_chunk2(File, Offset, Size, Depth, STime, TO,
+  when RestFLUs == [] orelse Depth == 0 ->
+       do_trim_midtail2(RestFLUs, Prefix, File, Offset, Size,
+                        Ws, Depth + 1, STime, TO, S);
+do_trim_midtail(_RestFLUs, Prefix, File, Offset, Size,
+                Ws, Depth, STime, TO, #state{proj=P}=S) ->
+    %% io:format(user, "midtail sleep2,", []),
+    sleep_a_while(Depth),
+    DiffMs = timer:now_diff(os:timestamp(), STime) div 1000,
+    if DiffMs > TO ->
+            {reply, {error, partition}, S};
+       true ->
+            S2 = update_proj(S#state{proj=undefined, bad_proj=P}),
+            case S2#state.proj of
+                undefined ->
+                    {reply, {error, partition}, S};
+                P2 ->
+                    RestFLUs2 = mutation_flus(P2),
+                    case RestFLUs2 -- Ws of
+                        RestFLUs2 ->
+                            %% None of the writes that we have done so far
+                            %% are to FLUs that are in the RestFLUs2 list.
+                            %% We are pessimistic here and assume that
+                            %% those FLUs are permanently dead.  Start
+                            %% over with a new sequencer assignment, at
+                            %% the 2nd have of the impl (we have already
+                            %% slept & refreshed the projection).
+
+                            if Prefix == undefined -> % atom! not binary()!!
+                                    {error, partition};
+                               true ->
+                                    do_trim_chunk(Prefix, Offset, Size,
+                                                  Depth, STime, TO, S2)
+                            end;
+                        RestFLUs3 ->
+                            do_trim_midtail2(RestFLUs3, Prefix, File, Offset, Size,
+                                             Ws, Depth + 1, STime, TO, S2)
+                    end
+            end
+    end.            
+
+do_trim_midtail2([], _Prefix, _File, _Offset, _Size,
+                   _Ws, _Depth, _STime, _TO, S) ->
+    %% io:format(user, "ok!\n", []),
+    {reply, ok, S};
+do_trim_midtail2([FLU|RestFLUs]=FLUs, Prefix, File, Offset, Size,
+                   Ws, Depth, STime, TO,
+                   #state{epoch_id=EpochID, proxies_dict=PD}=S) ->
+    Proxy = orddict:fetch(FLU, PD),
+    case ?FLU_PC:trim_chunk(Proxy, EpochID, File, Offset, Size, ?TIMEOUT) of
+        ok ->
+            %% io:format(user, "write ~w,", [FLU]),
+            do_trim_midtail2(RestFLUs, Prefix, File, Offset, Size,
+                             [FLU|Ws], Depth, STime, TO, S);
+        {error, bad_checksum}=BadCS ->
+            %% TODO: alternate strategy?
+            {reply, BadCS, S};
+        {error, Retry}
+          when Retry == partition; Retry == bad_epoch; Retry == wedged ->
+            do_trim_midtail(FLUs, Prefix, File, Offset, Size,
+                            Ws, Depth, STime, TO, S)
     end.
```

Comment:
Also, please check to see if there is any additional code required to handle the case of a racing write (by a regular client or a repair client) and a trim.